flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/SplitPythonConditionFromJoinRule.java

```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        FlinkLogicalJoin join = call.rel(0);
+        JoinRelType joinType = join.getJoinType();
+        // matches if it is inner join and it contains Python functions in condition
+        return joinType == JoinRelType.INNER
+                && Optional.ofNullable(join.getCondition())
+                        .map(rexNode -> containsPythonCall(rexNode, null))
+                        .orElse(false);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        FlinkLogicalJoin join = call.rel(0);
+        RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+
+        List<RexNode> joinFilters = RelOptUtil.conjunctions(join.getCondition());
+        List<RexNode> pythonFilters =
+                joinFilters.stream()
+                        .filter(rexNode -> containsPythonCall(rexNode, null))
+                        .collect(Collectors.toList());
+        List<RexNode> remainingFilters =
+                joinFilters.stream()
+                        .filter(rexNode -> !containsPythonCall(rexNode, null))
+                        .collect(Collectors.toList());
+
+        RexNode newJoinCondition = RexUtil.composeConjunction(rexBuilder, remainingFilters);
+        FlinkLogicalJoin bottomJoin =
+                new FlinkLogicalJoin(
+                        join.getCluster(),
+                        join.getTraitSet(),
+                        join.getLeft(),
+                        join.getRight(),
+                        newJoinCondition,
+                        join.getHints(),
+                        join.getJoinType());
+
+        RexProgram rexProgram =
+                new RexProgramBuilder(bottomJoin.getRowType(), rexBuilder).getProgram();
+        RexNode topCalcCondition = RexUtil.composeConjunction(rexBuilder, pythonFilters);
+
+        FlinkLogicalCalc topCalc =
+                new FlinkLogicalCalc(
+                        join.getCluster(),
+                        join.getTraitSet(),
+                        bottomJoin,
+                        RexProgram.create(
+                                bottomJoin.getRowType(),
+                                rexProgram.getExprList(),
+                                topCalcCondition,
+                                bottomJoin.getRowType(),
+                                rexBuilder));
+
+        call.transformTo(topCalc);
+    }
+
+    /** Rule configuration. */
+    @Value.Immutable(singleton = false)
+    public interface SplitPythonConditionFromJoinRuleConfig extends RelRule.Config {
+        SplitPythonConditionFromJoinRuleConfig DEFAULT =
+                ImmutableSplitPythonConditionFromJoinRule.SplitPythonConditionFromJoinRuleConfig
+                        .builder()
+                        .build()
+                        .withOperandSupplier(b0 -> b0.operand(FlinkLogicalJoin.class).noInputs())
+                        .withRelBuilderFactory(RelFactories.LOGICAL_BUILDER)
```

Comment:
we may not need this withRelBuilderFactory(RelFactories.LOGICAL_BUILDER) just for align